src/System.IO.FileSystem/src/System/IO/Win32FileStreamAsyncResult.cs

```diff
@@ -1,92 +1,49 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-using System;
-using Microsoft.Win32;
 using Microsoft.Win32.SafeHandles;
 using System.Security;
 using System.Threading;
 using System.Threading.Tasks;
 using System.Runtime.InteropServices;
-using System.Runtime.CompilerServices;
-using System.Globalization;
-using System.Runtime.Versioning;
 using System.Diagnostics;
 
 namespace System.IO
 {
     internal partial class Win32FileStream
     {
+#if USE_OVERLAPPED
         // This is an internal object implementing IAsyncResult with fields
         // for all of the relevant data necessary to complete the IO operation.
         // This is used by AsyncFSCallback and all of the async methods.
         unsafe private sealed class FileStreamAsyncResult : IAsyncResult
         {
-            // User code callback
-            private AsyncCallback _userCallback;
-            private Object _userStateObject;
-            private ManualResetEvent _waitHandle;
-#if USE_OVERLAPPED
-            private SafeFileHandle _handle;      // For cancellation support.
-            private NativeOverlapped* _overlapped;
-            internal NativeOverlapped* OverLapped {[SecurityCritical]get { return _overlapped; } }
-            internal bool IsAsync {[SecuritySafeCritical]get { return _overlapped != null; } }
-
-            internal int _EndXxxCalled;   // Whether we've called EndXxx already.
-            private int _numBytes;     // number of bytes read OR written
-
-            private int _errorCode;
-            internal int ErrorCode { get { return _errorCode; } }
-#endif
-
-            private int _numBufferedBytes;
+            private readonly TaskCompletionSource<int> _completionSource;
```

Comment:
You can have FileStreamAsyncResult (which should probably be renamed) derive from TaskCompletionSource<int> rather than containing a field of it.  In doing so, you save a field on the instance, and this type basically just becomes a TaskCompletionSource<int> that carries some additional state with it.